server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,911 @@
+
+		return this.query({
+			statement: 'deleteRequest', data: [senderID, receiverID], type: 'run',
+		});
+	}
+	async approveRequest(receiver: User, senderName: string) {
+		const senderID = toID(senderName);
+		return this.query({type: 'transaction', data: [senderID, receiver.id], statement: 'accept'});
+	}
+	async visualizeList(user: User) {
+		const friends = await this.getFriends(user, {includeLogin: true});
+		if (!friends.size) {
+			return `<h3>Your friends:</h3> <h4>None.</h4>`;
+		}
+		const categorized: {[k: string]: string[]} = {
+			online: [],
+			idle: [],
+			busy: [],
+			offline: [],
+		};
+		const loginTimes: {[k: string]: number} = {};
+		for (const {friend: friendName, login} of [...friends].sort() as AnyObject[]) {
+			const friend = Users.get(friendName as string);
+			if (friend?.connected) {
+				categorized[friend.statusType].push(friend.id);
+			} else {
+				categorized.offline.push(toID(friendName));
+				loginTimes[toID(friendName)] = login;
+			}
+		}
+
+		const sorted = Object.keys(categorized)
+			.filter(item => categorized[item].length > 0)
+			.map(item => `${STATUS_TITLES[item]} (${categorized[item].length})`);
+
+		let buf = `<h3>Your friends: <small> `;
+		if (sorted.length > 0) {
+			buf += `Total (${friends.size}) | ${sorted.join(' | ')}`;
+		} else {
+			buf += `</h3><em>you have no friends added on Showdown lol</em><br /><br /><br />`;
+			buf += `<strong>To add a friend, use </strong><code>/friend add [username]</code>.<br /><br />`;
+			buf += `<strong>To move over your friends to this account from a different account, `;
+			buf += `sign into that account and use </strong><code>/friend requesttransfer [new name]</code>.`;
+			return buf;
+		}
+		buf += `</h3> `;
+
+		for (const key in categorized) {
+			const friendArray = categorized[key].sort();
+			if (friendArray.length === 0) continue;
+			buf += `<h4>${STATUS_TITLES[key]} (${friendArray.length})</h4>`;
+			for (const friend of friendArray) {
+				buf += `<div class="pad"><div>`;
+				buf += this.displayFriend(friend, loginTimes[toID(friend)]);
+				buf += `</div></div>`;
+			}
+		}
+
+		return buf;
+	}
+	displayFriend(userid: string, login?: number) {
+		const user = Users.getExact(userid); // we want this to be exact
+		const connected = user?.connected;
+		const name = Utils.escapeHTML(user ? user.name : userid);
+		const statusType = connected ?
+			`<strong style="color:${STATUS_COLORS[user!.statusType]}">\u25C9 ${STATUS_TITLES[user!.statusType]}</strong>` :
+			'\u25CC Offline';
+		let buf = connected ?
+			`<span class="username"> <strong>${name}</strong></span><span><small> (${statusType})</small></span>` :
+			Utils.html`<i>${name}</i> <small>(${statusType})</small>`;
+		buf += `<br />`;
+
+		const oldName = this.renames.get(userid);
+		if (oldName) {
+			buf += Utils.html`<small>(recently renamed from ${oldName})</small><br />`;
+		}
+
+		const curUser = Users.get(userid); // might be an alt
+		if (user?.connected) {
```

Comment:
You assign `user?.connected` to the variable `connected` earlier, pick one or the other